File: examples/chapter02/matrixnorms.m
%%
A = [ 2  0; 1  -1 ]

%%
% The default norm returned by the |norm| command is the 2-norm.
twonorm = norm(A)

%%
% You can get the 1-norm as well.
onenorm = norm(A,1)

%%
% The 1-norm is equivalent to 
max( sum(abs(A),1) )   % sum along rows (1st matrix dimension)

%%
% Similarly, we can get the $\infty$-norm and check our formula for it.
infnorm = norm(A,inf)
max( sum(abs(A),2) )   % sum along columns (2nd matrix dimension)


%%
% Here we illustrate the geometric interpretation of the 2-norm. First, we
% will sample a lot of vectors on the unit circle in $\mathbf{R}^2$. 
theta = linspace(0,2*pi,601);
x = [ cos(theta); sin(theta) ];  % 601 unit columns
subplot(1,2,1), plot(x(1,:),x(2,:)), axis equal
title('Unit circle in 2-norm')
xlabel('x_1'),  ylabel('x_2')

%%
% We can apply |A| to every column of |x| simply by using
Ax = A*x;

%%
% We superimpose the image of the unit circle with the circle whose radius
% is $\|\mathbf{A}\|_2$, and display multiple plots
% with the @glsbegin@subplot@glsend@ command.
subplot(1,2,2), plot(Ax(1,:),Ax(2,:)), axis equal
hold on, plot(twonorm*x(1,:),twonorm*x(2,:),'--')
title('Image of Ax, showing ||A||')
xlabel('x_1'),  ylabel('x_2')

File: julia/scripts/chapter02/matrixnorms.jl

A = [ 2  0; 1  -1 ]

using LinearAlgebra
twonorm = opnorm(A)

onenorm = opnorm(A,1)

maximum( sum(abs.(A),dims=1) )   # sum down the rows (1st matrix dimension)

infnorm = opnorm(A,Inf)

maximum( sum(abs.(A),dims=2) )   # sum across columns (2nd matrix dimension)

theta = 2*pi*(0:1/600:1)
x = @. [ cos(theta) sin(theta) ]'   # 601 unit columns

using Plots
plot(x[1,:],x[2,:],aspect_ratio=1,layout=(1,2),subplot=1,
    title="Unit circle",leg=:none,xlabel="\$x_1\$",ylabel="\$x_2\$")

Ax = A*x;

plot!(Ax[1,:],Ax[2,:],subplot=2,aspect_ratio=1,
    title="Image under map",leg=:none,xlabel="\$x_1\$",ylabel="\$x_2\$")
plot!(twonorm*x[1,:],twonorm*x[2,:],subplot=2,l=:dash)

